### ISSUE TITLE
When in cluster registry is used, correct namespace should be used

### ISSUE THREAD

**Issue by zroubalik** [knative/func#1029]
Suppose I'd like to use in cluster registry, then if I do:
1. I am logged to namespace `foo`
2. `$ func deploy`, image is stored under `foo`
3. Set context to namespace `bar`
4. `$ func deploy` -> image still points to namespace `foo`, thus I am not able to deploy the function.
```

**Comment by matejvasek** [comment]
What is error? Failure to pull?

**Comment by zroubalik** [comment]
Yeah, no permissions to access the image (the namespace is part of the image name).

**Comment by zroubalik** [comment]
```
Failed to pull image "image-registry.openshift-image-registry.svc:5000/demo/hello-knativecon@sha256:27986272c2605d6989babb70074e3b34f564b95f76e0e6f3ae0165b9e4797ceb": rpc error: code = Unknown desc = reading manifest sha256:27986272c2605d6989babb70074e3b34f564b95f76e0e6f3ae0165b9e4797ceb in image-registry.openshift-image-registry.svc:5000/demo/hello-knativecon: unauthorized: authentication required
```

**Comment by zroubalik** [comment]
@matejvasek seems like, this is only related to on-cluster build. Because the KService is being deployed with permissions set to the ServiceAccount that belongs to the particular namespace.

**Comment by matejvasek** [comment]
@zroubalik  so when deploying from developer machine you can deploy service/image cross-namespace?

**Comment by zroubalik** [comment]
I think that it works because you are trying access namespaces that you have permissions to do so.

**Comment by matejvasek** [comment]
It is issue even for local development not only in-cluster-build.

>Though I think that we should anyway always reference (in the image) the current namespace.

Perhaps, but that would require quite a rewrite since image is set at `func create` and persisted in `func.yaml`.

**Comment by zroubalik** [comment]
> Perhaps, but that would require quite a rewrite since image is set at `func create` and persisted in `func.yaml`.

That's not always the case, it get's populated during build/deploy.

**Comment by lkingland** [comment]
Namespace calculation was refactored, and now has tests, which confirm that it correctly:
- Records in func.yaml the namespace in which the function is deployed
- Uses this namespace when redeploying by default
- Issues a warning when the user's currently active namespace is different than the function's current namespace
Should help avoid "Wher's my function?  I just deployed it!"
- Issues a warning when the requested deployment namespace differs from the function's current namespace
Should help avoid "Why do I have more than one of my function running in different namespaces"

I believe that addresses the (non-authorization-related) parts of this issue?

**Comment by gauron99** [comment]
Id like to work on this, id get to this after #1913 
/assign

**Comment by gauron99** [comment]
deploying in ns `ns-one` -> changing context to `ns-two` and deploying again with default config prints out a warning that function is deployed in `ns-one` and is continuing to work in said namespace. In this case `ns-two` is ignored.
 
If user adds a `--namespace ns-two` flag, the Function will be deployed in the `ns-two` namespace (overriding the namespace config in func.yaml) as well resulting in 2 Functions in cluster.

The original issue seems fine, the second issue will be addressed in separate issue
